File: src/systemd.rs
use std::path::PathBuf;
use super::genesis::*;
use std::fs::File;
use std::io::{Read, Write};
use std::process::Command;
use super::genesis;
use std::io::{Error,ErrorKind};

pub struct Emitter{
    prefix: PathBuf,
}

impl Emitter {
    pub fn new(prefix: PathBuf) -> Self {
        Self {
            prefix
        }
    }


    pub fn load(&mut self, config: &genesis::Genesis) -> Result<(), Error> {
        for (name, i) in &config.interface {
            println!("{:#?}", i);
            self.interface(name, i)?;
        }
        Command::new("systemctl")
            .args(&["restart", "systemd-networkd"])
            .status()?;
        Ok(())
    }
    pub fn commit(&self) -> Result<(), Error>  {
        Ok(())
    }

    fn interface(&self, name: &str, config: &Interface) -> Result<(), Error> {
        let p = self.prefix.join("etc/systemd/network/");
        std::fs::create_dir_all(&p)?;
        let p = p.join(format!("20-genesis-{}.network", name));
        let mut f = File::create(&p)?;

        let device = match &config.device {
            Some(v) => v,
            None => {
                return Err(Error::new(ErrorKind::Other, format!("interface.device is required")));
            }
        };

        write!(f, "[Match]\nName={}\n\n", device).unwrap();

        let mut dhcp = match &config.dhcp {
            Some(toml::value::Value::String(s)) =>  {
                match s.as_str() {
                    "yes" | "true" => {
                        Some("true".to_string())
                    },
                    "ipv6" => {
                        Some("ipv6".to_string())
                    },
                    "ipv4" => {
                        Some("ipv4".to_string())
                    },
                    "false" | "no" => {
                        None
                    },
                    _ => None,
                }
            },
            Some(toml::value::Value::Boolean(v)) => {
                if *v {
                    Some("true".to_string())
                } else {
                    None
                }
            },
            _ => None
        };

        if let Some(s) = &config.proto {
            if s == "dhcp" {
                dhcp = Some("ipv4".to_string())
            }
        }

        if let Some(dhcp) = dhcp {
            write!(f, "[Network]\nDHCP={}\n", dhcp)?;
            write!(f, "[DHCP]\nRouteMetric=10\n")?;
        } else {
            write!(f, "[Network]\n")?;
            if let Some(ipaddr) = &config.ipaddr {
                for ipaddr in ipaddr {
                    write!(f, "Address={}\n", ipaddr)?;
                }
            } else {
                return Err(Error::new(ErrorKind::Other, format!("either interface.dhcp or ipaddr required")));
            }
            if let Some(gw) = &config.gateway {
                write!(f, "Gateway={}\n", gw)?;
            }
            if let Some(dns) = &config.dns {
                for dns in dns {
                    write!(f, "DNS={}\n", dns)?;
                }
            }
        }


        if let Some(wifi) = &config.wifi {
            match wifi.mode{
                WifiMode::Ap | WifiMode::Monitor => {
                    return Err(Error::new(ErrorKind::Other, format!("wifi mode not supported")));
                }
                WifiMode::Sta=> {
                    let p = self.prefix.join("etc/wpa_supplicant/");
                    std::fs::remove_dir_all(&p).ok();
                    std::fs::create_dir_all(&p)?;
                    let p = p.join(format!("wpa_supplicant-{}.conf", device ));
                    let mut f = File::create(&p)?;
                    write!(f, "
                        ctrl_interface=/var/run/wpa_supplicant
                        ap_scan=1
                    ")?;
                    for sta in &config.sta {
                        match (&sta.auth, &sta.key) {
                            (Some(WifiAuth::None), _) | (None, _) | (_, None) => {
                                write!(f, "
                                network={{
                                    ssid=\"{}\"
                                    key_mgmt=NONE
                                }}
                                ", sta.ssid)?;
                            }
                            (Some(WifiAuth::Psk2), Some(key)) => {
                                write!(f, "
                                network={{
                                    key_mgmt=WPA-PSK
                                    ssid=\"{}\"
                                    psk=\"{}\"
                                }}
                                ", sta.ssid, key)?;
                            }
                        }
                    }
                    drop(f);
                    Command::new("systemctl")
                        .args(&["start", &format!("wpa_supplicant@{}.service", device)])
                        .status()?;
                }
            }
        }

        Ok(())
    }
}

File: systemd.go
// +build systemd

package main

import (
	"errors"
	"fmt"
	"os"
	"path/filepath"
)

const backend = "systemd"

type emitter struct {
	prefix string
	files  []string
}

func newEmitter() emitter {
	return emitter{
		prefix: "./",
	}
}

func (se emitter) load(config genesis) error {
	for name, devIntf := range config.Interfaces {
		err := se.loadDeviceInterface(name, devIntf)
		if err != nil {
			return err
		}
	}
	err := compress("etc/config", se.files...)
	if err != nil {
		return err
	}

	return nil
}

func (se emitter) commit() error {
	return nil
}

func (se emitter) loadDeviceInterface(name string, devIntf devInterface) error {
	path := filepath.Join(se.prefix, "etc/systemd/network/")
	err := os.MkdirAll(path, 0700)
	if err != nil {
		return err
	}
	path = filepath.Join(path, fmt.Sprintf("20-genesis-%s.network", name))
	file, err := os.Create(path)
	se.files = append(se.files, path)
	defer file.Close()
	if err != nil {
		return err
	}
	if devIntf.Device == "" {
		return errors.New("interface.device is required")
	}
	_, err = fmt.Fprintf(file, "[Match]\nName=%s\n\n", devIntf.Device)
	if err != nil {
		return err
	}

	var dhcpSetting string
	switch dhcp := devIntf.Dhcp.(type) {
	case string:
		switch dhcp {
		case "yes", "true":
			dhcpSetting = "true"
		case "no", "false":
			break
		case "ipv4":
			dhcpSetting = "ipv4"
		case "ipv6":
			dhcpSetting = "ipv6"
		}
	case bool:
		if dhcp {
			dhcpSetting = "true"
		}
	}
	if devIntf.Proto == "dhcp" {
		dhcpSetting = "ipv4"
	}
	if dhcpSetting != "" {
		_, err = fmt.Fprintf(file, `[Network]
DHCP=%s
[DHCP]
RouteMetric=10
`, dhcpSetting)
		if err != nil {
			return err
		}

	} else {
		_, err = fmt.Fprintf(file, "[Network]\n")
		if err != nil {
			return err
		}
		if len(devIntf.IPAddres) == 0 {
			return errors.New("either interface.dhcp or ipaddr required")
		}

		for _, ip := range devIntf.IPAddres {
			_, err = fmt.Fprintf(file, "Address=%s\n", ip)
			if err != nil {
				return err
			}
		}
		if devIntf.Gateway != "" {
			_, err = fmt.Fprintf(file, "Gateway=%s\n", devIntf.Gateway)
			if err != nil {
				return err
			}
		}
		for _, dns := range devIntf.DNS {
			_, err = fmt.Fprintf(file, "DNS=%s\n", dns)
			if err != nil {
				return err
			}
		}
	}
	switch devIntf.Wifi.Mode {
	case sta:
		path := filepath.Join(se.prefix, "etc/systemd/network/")
		os.RemoveAll(path)
		os.MkdirAll(path, 0700)
		path = filepath.Join(path, fmt.Sprintf("wpa_supplicant-%s.conf", devIntf.Device))
		file2, err := os.Create(path)
		se.files = append(se.files, path)
		defer file2.Close()
		if err != nil {
			return err
		}
		_, err = fmt.Fprintf(file2, `
ctrl_interface=/var/run/wpa_supplicant
ap_scan=1
`)
		if err != nil {
			return err
		}
		for _, sta := range devIntf.STA {
			if sta.Auth == psk2 && sta.Key != "" {
				_, err = fmt.Fprintf(file2, `
network={{
	key_mgmt=WPA-PSK
	ssid="%s"
	psk="%s"
}}
`, sta.Ssid, sta.Key)
				if err != nil {
					return err
				}
			} else {
				_, err = fmt.Fprintf(file2, `
network={{
	ssid="%s"
	key_mgmt=NONE
}}
`, sta.Ssid)
				if err != nil {
					return err
				}
			}
		}
	case ap, monitor:
		return errors.New("wifi mode not supported")
	}
	return nil
}
